s3torchconnector/src/s3torchconnector/dcp/fsdp_filesystem.py

```diff
@@ -0,0 +1,249 @@
+
+        Returns:
+            Union[str, os.PathLike]: The initialized path.
+        """
+        logger.debug("init_path for %s", path)
+        self._path = path
+        return self._path
+
+    def rename(
+        self, old_path: Union[str, os.PathLike], new_path: Union[str, os.PathLike]
+    ) -> None:
+        """Rename an object in S3.
+
+        This is emulated by copying it to a new path and deleting the old path. The deletion part is retried (see also
+        :func:`S3FileSystem._delete_with_retry`).
+
+        Args:
+            old_path (Union[str, os.PathLike]): The current path of the object.
+            new_path (Union[str, os.PathLike]): The new path for the object.
+
+        Raises:
+            ValueError: If the old and new paths point to different buckets.
+            S3Exception: If there is an error with the S3 client.
+        """
+        logger.debug("rename %s to %s", old_path, new_path)
+
+        old_path = _path_to_str(old_path)
+        new_path = _path_to_str(new_path)
+
+        old_bucket, old_key = parse_s3_uri(old_path)
+        new_bucket, new_key = parse_s3_uri(new_path)
+
+        if old_bucket != new_bucket:
+            raise ValueError(
+                f"Source and destination buckets cannot be different (rename does not support cross-buckets operations)"
+            )
+
+        self._client.copy_object(
+            src_bucket=old_bucket,
+            src_key=old_key,
+            dst_bucket=new_bucket,
+            dst_key=new_key,
+        )
+        self._delete_with_retry(old_bucket, old_key)
+
+    def mkdir(self, path: Union[str, os.PathLike]) -> None:
+        """No-op method for creating directories in S3 (not needed)."""
+        pass
+
+    def exists(self, path: Union[str, os.PathLike]) -> bool:
+        logger.debug("exists %s", path)
+
+        path = _path_to_str(path)
+        bucket, key = parse_s3_uri(path)
+        try:
+            self._client.head_object(bucket, key)
+        except S3Exception:
+            return False
+        return True
+
+    def rm_file(self, path: Union[str, os.PathLike]) -> None:
+        logger.debug("remove %s", path)
+
+        path = _path_to_str(path)
+        bucket, key = parse_s3_uri(path)
+        try:
+            self._client.delete_object(bucket, key)
+        except S3Exception:
+            logger.exception("Failed to remove object from S3")
+
+    @classmethod
+    def validate_checkpoint_id(cls, checkpoint_id: Union[str, os.PathLike]) -> bool:
+        logger.debug("validate_checkpoint_id for %s", checkpoint_id)
+
+        if isinstance(checkpoint_id, Path):
+            return True
+
+        try:
+            parse_s3_uri(_path_to_str(checkpoint_id))
+        except ValueError:
+            return False
+        return True
+
+    @retry(
+        retry=retry_if_exception_type(S3Exception),
+        stop=stop_after_attempt(3),
+        wait=wait_fixed(3) + wait_random(0, 2),
```

Comment:
should we go with exponential backoff strategy?

Reply:
Why not: would you recommend any parameters? Like, I can think of:
```
@retry(
    retry=retry_if_exception_type(S3Exception),
    stop=stop_after_attempt(3),
    wait=wait_random_exponential(multiplier=1, max=5),
    before_sleep=before_sleep_log(logger, logging.WARNING),
    after=after_log(logger, logging.ERROR),
    reraise=True,
)
```

Reply:
Also, should we do the retry for some specific scenarios. In this case, it looks like we are retrying for all errors, irrespective of whether it is retryable or not.

Reply:
That's true: I can think of:
1. [`Request throttled`](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3-client/src/s3_crt_client.rs#L1030-L1032)
2. [`Failed to construct request`](https://github.com/awslabs/mountpoint-s3/blob/main/mountpoint-s3-client/src/s3_crt_client.rs#L1006-L1008)

WDYT?